### ISSUE TITLE
Does blocking facebookexternalhit also break sharing to social media?

### ISSUE THREAD

**Issue by njt1982** [ai-robots-txt/ai.robots.txt#40]
https://github.com/ai-robots-txt/ai.robots.txt/blob/6b8d7f5890d6bed722a95297996c054c210bd3b8/robots.txt#L33

If we block this in robots.txt, will we affect the functionality for when URLs are shared to facebook from the site and Facebook sends that bot to get the Open Graph data for things like title and image for the post?

Ideally I'd like to block the bot from crawling / DoS'ing the site but still allow on-demand/cached page requests for OG data when a post is shared. Facebook does not need to crawl an entire site! :)

**Comment by cdransf** [comment]
It will, per Meta's docs:

> The primary purpose of FacebookExternalHit is to crawl the content of an app or website that was shared on one of Meta’s family of apps, such as Facebook, Instagram, or Messenger. 

I'm ok blocking Meta's products generally, but I'd weigh how important that functionality is against the traffic that you're seeing from that crawler.

**Comment by glyn** [comment]
> Facebook sends that bot to get the Open Graph data for things like title and image for the post

Although I don't use Facebook, I'd be surprised if sharing a link on Facebook required the crawler to run. Crawlers typically run on their own schedule rather than synchronously to a user action such as creating a post. Have you experimented by monitoring access of your website by such a crawler and then posting on Facebook and seeing if a "crawl" occurs before the post is available?

**Comment by paulrudy** [comment]
I've noticed that blocking `facebookexternalhit` prevents rich links (cards) from displaying in Apple Messages and Apple Mail (iOS and macOS). The change takes place immediately: blocking `facebookexternalhit` immediately prevents adding rich link, allowing immediately permits adding a rich link.

**Comment by glyn** [comment]
facebookexternalhit isn't in our list of AI crawlers, so isn't this discussion a bit moot?

**Comment by glyn** [comment]
Ah yes, you're right. I made the mistake of thinking the list was in
alphabetical order. Apologies.

**Comment by njt1982** [comment]
> Ah yes, you're right. I made the mistake of thinking the list was in alphabetical order. Apologies.

It's probably worth alphabetising them; as the list grows, duplicates are more likely.

Could be a github pre-commit command that sorts / uniques the list? 🤷🏻‍♂️

**Comment by glyn** [comment]
I would say the priority of this project is to block AI crawlers. It is not clear that facebookexternalhit gathers data for AI training, but we don't know it isn't either. I would personally vote to keep this in the list. The possible downsides seem negligible from my perspective. Any websites that really need not to block that user agent don't have to.

**Comment by njt1982** [comment]
This particular one is a tricky one as it is a very aggressive crawler... but, unlike some of them, a lot of our customers would likely notice if their website suddenly stopped displaying article image cards.

Maybe the solution here is a comment above it describing what it does and what the risks are?

Some of these will have a much lower risk profile... But on the flip side items like this (and potentially the Google ones, too) might have unintended impact for site SEO and Social exposure for those simply copy-pasting a list in to their site to try to stop these bots from taking down the server.

**Comment by glyn** [comment]
> Maybe the solution here is a comment above it describing what it does and what the risks are?

I like the sound of this. Would it be possible to demonstrate that these risks are real and not imaginary?

(I submitted a PR to extend the FAQ to take into account your "taking down the server" point - thanks!)

**Comment by mhattersley** [comment]
https://developers.facebook.com/docs/sharing/webmasters/web-crawlers/

**Comment by glyn** [comment]
That link was helpful.

Also Dark Visitors classifies this crawler as a "fetcher" which we [currently exclude](https://github.com/ai-robots-txt/ai.robots.txt/blob/71db599b41e476555e5d718942e7e8bfffdc30d6/code/dark_visitors.py#L35).

PR raised to fix this issue.

**Comment by jperezr21** [comment]
![Image](https://github.com/user-attachments/assets/3f82018d-f8ac-464d-b542-99d06d612590)

They're abusing this bot for other purposes. My website isn't shared *that* much.

**Comment by glyn** [comment]
> > Perhaps we should submit a new PR?
> 
> By all means. I noted in [#21](https://github.com/ai-robots-txt/ai.robots.txt/pull/21) that
> 
> > We also observed that it didn’t double-check robots.txt before starting another round of binge crawling.
> 
> So perhaps No for `respect`?

This issue is closed. Please raise a new issue or PR to capture any missing changes.

**Comment by paulrudy** [comment]
New PR: #154

**Comment by guest20** [comment]
Generating a preview for social media just needs the <head> section, so you can just stop sending the page once you're done with that section

You can likely hack up your app to early return when you have enough for the <meta> parts...

Or you can hang up when answering from cache _if u nasty_

**Comment by paulrudy** [comment]
@guest20 thanks for the idea. It inspired me to serve an empty body to the bot for pages that already have OpenGraph metadata.
